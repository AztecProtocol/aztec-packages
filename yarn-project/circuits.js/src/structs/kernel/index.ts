export * from './private_to_rollup_accumulated_data.js';
export * from './combined_constant_data.js';
export * from './private_to_rollup_kernel_circuit_public_inputs.js';
export * from './private_accumulated_data.js';
export * from './private_call_data.js';
export * from './private_kernel_circuit_public_inputs.js';
export * from './private_kernel_data.js';
export * from './private_kernel_init_circuit_private_inputs.js';
export * from './private_kernel_inner_circuit_private_inputs.js';
export * from './private_kernel_reset_circuit_private_inputs.js';
export * from './private_kernel_reset_dimensions.js';
export * from './private_kernel_reset_hints.js';
export * from './private_kernel_tail_circuit_private_inputs.js';
export * from './private_kernel_tail_circuit_public_inputs.js';
export * from './private_to_avm_accumulated_data.js';
export * from './private_to_public_accumulated_data.js';
export * from './private_to_public_accumulated_data_builder.js';
export * from './private_to_public_kernel_circuit_public_inputs.js';
export * from './tx_constant_data.js';
