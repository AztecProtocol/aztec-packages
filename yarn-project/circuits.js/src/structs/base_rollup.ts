import { BufferReader } from '../utils/buffer_reader.js';
import { assertLength, FieldsOf } from '../utils/jsUtils.js';
import { serializeToBuffer } from '../utils/serialize.js';
import {
  CONTRACT_TREE_HEIGHT,
  CONTRACT_TREE_ROOTS_TREE_HEIGHT,
  KERNEL_NEW_NULLIFIERS_LENGTH,
  NULLIFIER_TREE_HEIGHT,
  PRIVATE_DATA_TREE_HEIGHT,
  PRIVATE_DATA_TREE_ROOTS_TREE_HEIGHT,
} from './constants.js';
import { PreviousKernelData } from './kernel.js';
import { AggregationObject, Fr, MembershipWitness, UInt32 } from './shared.js';

export class NullifierLeafPreimage {
  constructor(public leafValue: Fr, public nextValue: Fr, public nextIndex: UInt32) {}

  toBuffer() {
    return serializeToBuffer(this.leafValue, this.nextValue, this.nextIndex);
  }
}

export class AppendOnlyTreeSnapshot {
  constructor(public root: Fr, public nextAvailableLeafIndex: UInt32) {}

  toBuffer() {
    return serializeToBuffer(this.root, this.nextAvailableLeafIndex);
  }

  static fromBuffer(buffer: Buffer | BufferReader): AppendOnlyTreeSnapshot {
    const reader = BufferReader.asReader(buffer);
    return new AppendOnlyTreeSnapshot(reader.readFr(), reader.readNumber());
  }
}

export class ConstantBaseRollupData {
  constructor(
    // The very latest roots as at the very beginning of the entire rollup:
    public startTreeOfHistoricPrivateDataTreeRootsSnapshot: AppendOnlyTreeSnapshot,
    public startTreeOfHistoricContractTreeRootsSnapshot: AppendOnlyTreeSnapshot,
    public treeOfHistoricL1ToL2MsgTreeRootsSnapshot: AppendOnlyTreeSnapshot,

    // Some members of this struct tbd:
    public privateKernelVkTreeRoot: Fr,
    public publicKernelVkTreeRoot: Fr,
    public baseRollupVkHash: Fr,
    public mergeRollupVkHash: Fr,
  ) {}

  static from(fields: FieldsOf<ConstantBaseRollupData>): ConstantBaseRollupData {
    return new ConstantBaseRollupData(...ConstantBaseRollupData.getFields(fields));
  }

  static fromBuffer(buffer: Buffer | BufferReader): ConstantBaseRollupData {
    const reader = BufferReader.asReader(buffer);
    return new ConstantBaseRollupData(
      reader.readObject(AppendOnlyTreeSnapshot),
      reader.readObject(AppendOnlyTreeSnapshot),
      reader.readObject(AppendOnlyTreeSnapshot),
      reader.readFr(),
      reader.readFr(),
      reader.readFr(),
      reader.readFr(),
    );
  }

  static getFields(fields: FieldsOf<ConstantBaseRollupData>) {
    return [
      fields.startTreeOfHistoricPrivateDataTreeRootsSnapshot,
      fields.startTreeOfHistoricContractTreeRootsSnapshot,
      fields.treeOfHistoricL1ToL2MsgTreeRootsSnapshot,
      fields.privateKernelVkTreeRoot,
      fields.publicKernelVkTreeRoot,
      fields.baseRollupVkHash,
      fields.mergeRollupVkHash,
    ] as const;
  }

  toBuffer() {
    return serializeToBuffer(...ConstantBaseRollupData.getFields(this));
  }
}

/**
 * Inputs to the base rollup circuit
 */
export class BaseRollupInputs {
  constructor(
    public kernelData: [PreviousKernelData, PreviousKernelData],

    public startPrivateDateTreeSnapshot: AppendOnlyTreeSnapshot,
    public startNullifierTreeSnapshot: AppendOnlyTreeSnapshot,
    public startContractTreeSnapshot: AppendOnlyTreeSnapshot,

    public lowNullifierLeafPreimages: NullifierLeafPreimage[],
    public lowNullifierMembershipWitness: MembershipWitness<typeof NULLIFIER_TREE_HEIGHT>[],

    public newCommitmentsSubtreeSiblingPath: Fr[],
    public newNullifiersSubtreeSiblingPath: Fr[],
    public newContractsSubtreeSiblingPath: Fr[],

    public historicPrivateDataTreeRootMembershipWitnesses: [
      MembershipWitness<typeof PRIVATE_DATA_TREE_ROOTS_TREE_HEIGHT>,
      MembershipWitness<typeof PRIVATE_DATA_TREE_ROOTS_TREE_HEIGHT>,
    ],
    public historicContractsTreeRootMembershipWitnesses: [
      MembershipWitness<typeof CONTRACT_TREE_ROOTS_TREE_HEIGHT>,
      MembershipWitness<typeof CONTRACT_TREE_ROOTS_TREE_HEIGHT>,
    ],

    public constants: ConstantBaseRollupData,
  ) {
    assertLength(this, 'lowNullifierLeafPreimages', 2 * KERNEL_NEW_NULLIFIERS_LENGTH);
    assertLength(this, 'lowNullifierMembershipWitness', 2 * KERNEL_NEW_NULLIFIERS_LENGTH);
    assertLength(this, 'newCommitmentsSubtreeSiblingPath', PRIVATE_DATA_TREE_HEIGHT);
    assertLength(this, 'newNullifiersSubtreeSiblingPath', NULLIFIER_TREE_HEIGHT);
    assertLength(this, 'newContractsSubtreeSiblingPath', CONTRACT_TREE_HEIGHT);
  }

  static from(fields: FieldsOf<BaseRollupInputs>): BaseRollupInputs {
    return new BaseRollupInputs(...BaseRollupInputs.getFields(fields));
  }

  static getFields(fields: FieldsOf<BaseRollupInputs>) {
    return [
      fields.kernelData,
      fields.startPrivateDateTreeSnapshot,
      fields.startNullifierTreeSnapshot,
      fields.startContractTreeSnapshot,
      fields.lowNullifierLeafPreimages,
      fields.lowNullifierMembershipWitness,
      fields.newCommitmentsSubtreeSiblingPath,
      fields.newNullifiersSubtreeSiblingPath,
      fields.newContractsSubtreeSiblingPath,
      fields.historicPrivateDataTreeRootMembershipWitnesses,
      fields.historicContractsTreeRootMembershipWitnesses,
      fields.constants,
    ] as const;
  }

  toBuffer() {
    return serializeToBuffer(...BaseRollupInputs.getFields(this));
  }
}

/**
 * Output of the base rollup circuit
 */
export class BaseRollupPublicInputs {
  constructor(
    public endAggregationObject: AggregationObject,
    public constants: ConstantBaseRollupData,

    public startPrivateDataTreeSnapshot: AppendOnlyTreeSnapshot,
    public endPrivateDataTreeSnapshot: AppendOnlyTreeSnapshot,

    public startNullifierTreeSnapshot: AppendOnlyTreeSnapshot,
    public endNullifierTreeSnapshots: AppendOnlyTreeSnapshot,

    public startContractTreeSnapshot: AppendOnlyTreeSnapshot,
    public endContractTreeSnapshot: AppendOnlyTreeSnapshot,

    // Hashes (sha256), to make public inputs constant-sized (to then be unpacked on-chain). Length 2 for high and low
    public calldataHash: [Fr, Fr],
  ) {}

  /**
   * Deserializes from a buffer or reader, corresponding to a write in cpp.
   * @param bufferReader - Buffer to read from.
   */
  static fromBuffer(buffer: Buffer | BufferReader): BaseRollupPublicInputs {
    const reader = BufferReader.asReader(buffer);
    return new BaseRollupPublicInputs(
      reader.readObject(AggregationObject),
      reader.readObject(ConstantBaseRollupData),
      reader.readObject(AppendOnlyTreeSnapshot),
      reader.readObject(AppendOnlyTreeSnapshot),
      reader.readObject(AppendOnlyTreeSnapshot),
      reader.readObject(AppendOnlyTreeSnapshot),
      reader.readObject(AppendOnlyTreeSnapshot),
      reader.readObject(AppendOnlyTreeSnapshot),
      reader.readArray(2, Fr) as [Fr, Fr],
    );
  }

  /**
   * Serialize this as a buffer.
   * @returns The buffer.
   */
  toBuffer() {
    return serializeToBuffer(
      this.endAggregationObject,
      this.constants,

      this.startPrivateDataTreeSnapshot,
      this.endPrivateDataTreeSnapshot,

      this.startNullifierTreeSnapshot,
      this.endNullifierTreeSnapshots,

      this.startContractTreeSnapshot,
      this.endContractTreeSnapshot,

      this.calldataHash,
    );
  }
}
