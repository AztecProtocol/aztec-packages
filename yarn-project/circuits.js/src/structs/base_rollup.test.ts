import { expectReserializeToMatchObject, expectSerializeToMatchSnapshot } from '../tests/expectSerialize.js';
import {
  fr,
  makeAppendOnlyTreeSnapshot,
  makeBaseRollupPublicInputs,
  makeConstantBaseRollupData,
  makePreviousKernelData,
} from '../tests/factories.js';
import { range } from '../utils/jsUtils.js';
import { CircuitsWasm } from '../wasm/circuits_wasm.js';
import { BaseRollupInputs, BaseRollupPublicInputs, NullifierLeafPreimage } from './base_rollup.js';
import {
  CONTRACT_TREE_HEIGHT,
  CONTRACT_TREE_ROOTS_TREE_HEIGHT,
  KERNEL_NEW_NULLIFIERS_LENGTH,
  NULLIFIER_TREE_HEIGHT,
  PRIVATE_DATA_TREE_HEIGHT,
  PRIVATE_DATA_TREE_ROOTS_TREE_HEIGHT,
} from './constants.js';
import { PreviousKernelData } from './kernel.js';
import { MembershipWitness } from './shared.js';

describe('structs/base_rollup', () => {
  it(`serializes and prints BaseRollupInputs`, async () => {
    const kernelData: [PreviousKernelData, PreviousKernelData] = [
      makePreviousKernelData(0x100),
      makePreviousKernelData(0x200),
    ];

    const startPrivateDateTreeSnapshot = makeAppendOnlyTreeSnapshot(0x100);
    const startNullifierTreeSnapshot = makeAppendOnlyTreeSnapshot(0x200);
    const startContractTreeSnapshot = makeAppendOnlyTreeSnapshot(0x300);

    const lowNullifierLeafPreimages = range(2 * KERNEL_NEW_NULLIFIERS_LENGTH, 0x1000).map(
      x => new NullifierLeafPreimage(fr(x), fr(x + 0x100), x + 0x200),
    );

    const lowNullifierMembershipWitness = range(2 * KERNEL_NEW_NULLIFIERS_LENGTH, 0x2000).map(x =>
      MembershipWitness.mock(NULLIFIER_TREE_HEIGHT, x),
    );

    const newCommitmentsSubtreeSiblingPath = range(PRIVATE_DATA_TREE_HEIGHT, 0x3000).map(x => fr(x));
    const newNullifiersSubtreeSiblingPath = range(NULLIFIER_TREE_HEIGHT, 0x4000).map(x => fr(x));
    const newContractsSubtreeSiblingPath = range(CONTRACT_TREE_HEIGHT, 0x5000).map(x => fr(x));

    const historicPrivateDataTreeRootMembershipWitnesses: BaseRollupInputs['historicPrivateDataTreeRootMembershipWitnesses'] =
      [
        MembershipWitness.mock(PRIVATE_DATA_TREE_ROOTS_TREE_HEIGHT, 0x6000),
        MembershipWitness.mock(PRIVATE_DATA_TREE_ROOTS_TREE_HEIGHT, 0x7000),
      ];

    const historicContractsTreeRootMembershipWitnesses: BaseRollupInputs['historicContractsTreeRootMembershipWitnesses'] =
      [
        MembershipWitness.mock(CONTRACT_TREE_ROOTS_TREE_HEIGHT, 0x8000),
        MembershipWitness.mock(CONTRACT_TREE_ROOTS_TREE_HEIGHT, 0x9000),
      ];

    const constants = makeConstantBaseRollupData(0x100);

    const baseRollupInputs = BaseRollupInputs.from({
      kernelData,
      startPrivateDateTreeSnapshot,
      startNullifierTreeSnapshot,
      startContractTreeSnapshot,
      lowNullifierLeafPreimages,
      lowNullifierMembershipWitness,
      newCommitmentsSubtreeSiblingPath,
      newNullifiersSubtreeSiblingPath,
      newContractsSubtreeSiblingPath,
      historicPrivateDataTreeRootMembershipWitnesses,
      historicContractsTreeRootMembershipWitnesses,
      constants,
    });

    const wasm = await CircuitsWasm.new();
    await expectSerializeToMatchSnapshot(
      baseRollupInputs.toBuffer(),
      'abis__test_roundtrip_serialize_base_rollup_inputs',
      wasm,
    );
  });

  it(`serializes and prints BaseRollupPublicInputs`, async () => {
    const baseRollupPublicInputs = makeBaseRollupPublicInputs();

    await expectSerializeToMatchSnapshot(
      baseRollupPublicInputs.toBuffer(),
      'abis__test_roundtrip_serialize_base_rollup_public_inputs',
    );
  });

  it(`serializes and deserializes BaseRollupPublicInputs`, async () => {
    const baseRollupPublicInputs = makeBaseRollupPublicInputs();

    await expectReserializeToMatchObject(
      baseRollupPublicInputs,
      'abis__test_roundtrip_reserialize_base_rollup_public_inputs',
      BaseRollupPublicInputs.fromBuffer,
    );
  });
});
