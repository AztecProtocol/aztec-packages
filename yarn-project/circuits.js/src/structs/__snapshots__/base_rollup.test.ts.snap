// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`structs/base_rollup serializes and prints BaseRollupInputs 1`] = `
"kernel_data:
[ public_inputs: end:
aggregation_object:
P0: { 0x100, 0x101 }
P1: { 0x200, 0x201 }
public_inputs: [
  0: 0x102
  1: 0x103
  2: 0x104
  3: 0x105
]
proof_witness_indices: [
  0: 106
  1: 107
  2: 108
  3: 109
  4: 10a
  5: 10b
has_data: 0

private_call_count: 0x10c
new_commitments:
[ 0x200 0x201 0x202 0x203 0x204 0x205 0x206 0x207 0x208 0x209 0x20a 0x20b 0x20c 0x20d 0x20e 0x20f ]
new_nullifiers:
[ 0x300 0x301 0x302 0x303 0x304 0x305 0x306 0x307 0x308 0x309 0x30a 0x30b 0x30c 0x30d 0x30e 0x30f ]
private_call_stack:
[ 0x400 0x401 0x402 0x403 0x404 0x405 0x406 0x407 ]
public_call_stack:
[ 0x500 0x501 0x502 0x503 0x504 0x505 0x506 0x507 ]
l1_msg_stack:
[ 0x600 0x601 0x602 0x603 ]
new_contracts:
[ contract_address: 0x700
portal_contract_address: 0x101010101010101010101010101010101010101
function_tree_root: 0x702
 contract_address: 0x701
portal_contract_address: 0x202020202020202020202020202020202020202
function_tree_root: 0x703
 contract_address: 0x702
portal_contract_address: 0x303030303030303030303030303030303030303
function_tree_root: 0x704
 contract_address: 0x703
portal_contract_address: 0x404040404040404040404040404040404040404
function_tree_root: 0x705
 contract_address: 0x704
portal_contract_address: 0x505050505050505050505050505050505050505
function_tree_root: 0x706
 contract_address: 0x705
portal_contract_address: 0x606060606060606060606060606060606060606
function_tree_root: 0x707
 contract_address: 0x706
portal_contract_address: 0x707070707070707070707070707070707070707
function_tree_root: 0x708
 contract_address: 0x707
portal_contract_address: 0x808080808080808080808080808080808080808
function_tree_root: 0x709
 ]
optionally_revealed_data:
[ call_stack_item_hash: 0x800
function_data:
function_selector: 801
is_private: 1
is_constructor: 1

emitted_events:
[ 0x900 0x901 0x902 0x903 ]
vk_hash: 0x802
portal_contract_address: 0x303030303030303030303030303030303030303
pay_fee_from_l1: 1
pay_fee_from_public_l2: 0
called_from_l1: 1
called_from_public_l2: 0
 call_stack_item_hash: 0x801
function_data:
function_selector: 802
is_private: 1
is_constructor: 1

emitted_events:
[ 0x901 0x902 0x903 0x904 ]
vk_hash: 0x803
portal_contract_address: 0x404040404040404040404040404040404040404
pay_fee_from_l1: 1
pay_fee_from_public_l2: 0
called_from_l1: 1
called_from_public_l2: 0
 call_stack_item_hash: 0x802
function_data:
function_selector: 803
is_private: 1
is_constructor: 1

emitted_events:
[ 0x902 0x903 0x904 0x905 ]
vk_hash: 0x804
portal_contract_address: 0x505050505050505050505050505050505050505
pay_fee_from_l1: 1
pay_fee_from_public_l2: 0
called_from_l1: 1
called_from_public_l2: 0
 call_stack_item_hash: 0x803
function_data:
function_selector: 804
is_private: 1
is_constructor: 1

emitted_events:
[ 0x903 0x904 0x905 0x906 ]
vk_hash: 0x805
portal_contract_address: 0x606060606060606060606060606060606060606
pay_fee_from_l1: 1
pay_fee_from_public_l2: 0
called_from_l1: 1
called_from_public_l2: 0
 ]

constants:
old_tree_roots: private_data_tree_root: 0x200
nullifier_tree_root: 0x201
contract_tree_root: 0x202
private_kernel_vk_tree_root: 0x203

tx_context: is_fee_payment_tx: 0
is_rebate_payment_tx: 0
is_contract_deployment_tx: 1
contract_deployment_data: 
constructor_vk_hash: 0x204
function_tree_root: 0x205
contract_address_salt: 0x206
portal_contract_address: 0x707070707070707070707070707070707070707


is_private: 1

proof: [ 80 80 80 80 80 80 80 80 80 80 80 80 80 80 80 80 ]
vk: 0x12e040
vk_index: 42
vk_path: [ 0x1000 0x1001 0x1002 ]
 public_inputs: end:
aggregation_object:
P0: { 0x200, 0x201 }
P1: { 0x300, 0x301 }
public_inputs: [
  0: 0x202
  1: 0x203
  2: 0x204
  3: 0x205
]
proof_witness_indices: [
  0: 206
  1: 207
  2: 208
  3: 209
  4: 20a
  5: 20b
has_data: 0

private_call_count: 0x20c
new_commitments:
[ 0x300 0x301 0x302 0x303 0x304 0x305 0x306 0x307 0x308 0x309 0x30a 0x30b 0x30c 0x30d 0x30e 0x30f ]
new_nullifiers:
[ 0x400 0x401 0x402 0x403 0x404 0x405 0x406 0x407 0x408 0x409 0x40a 0x40b 0x40c 0x40d 0x40e 0x40f ]
private_call_stack:
[ 0x500 0x501 0x502 0x503 0x504 0x505 0x506 0x507 ]
public_call_stack:
[ 0x600 0x601 0x602 0x603 0x604 0x605 0x606 0x607 ]
l1_msg_stack:
[ 0x700 0x701 0x702 0x703 ]
new_contracts:
[ contract_address: 0x800
portal_contract_address: 0x101010101010101010101010101010101010101
function_tree_root: 0x802
 contract_address: 0x801
portal_contract_address: 0x202020202020202020202020202020202020202
function_tree_root: 0x803
 contract_address: 0x802
portal_contract_address: 0x303030303030303030303030303030303030303
function_tree_root: 0x804
 contract_address: 0x803
portal_contract_address: 0x404040404040404040404040404040404040404
function_tree_root: 0x805
 contract_address: 0x804
portal_contract_address: 0x505050505050505050505050505050505050505
function_tree_root: 0x806
 contract_address: 0x805
portal_contract_address: 0x606060606060606060606060606060606060606
function_tree_root: 0x807
 contract_address: 0x806
portal_contract_address: 0x707070707070707070707070707070707070707
function_tree_root: 0x808
 contract_address: 0x807
portal_contract_address: 0x808080808080808080808080808080808080808
function_tree_root: 0x809
 ]
optionally_revealed_data:
[ call_stack_item_hash: 0x900
function_data:
function_selector: 901
is_private: 1
is_constructor: 1

emitted_events:
[ 0xa00 0xa01 0xa02 0xa03 ]
vk_hash: 0x902
portal_contract_address: 0x303030303030303030303030303030303030303
pay_fee_from_l1: 1
pay_fee_from_public_l2: 0
called_from_l1: 1
called_from_public_l2: 0
 call_stack_item_hash: 0x901
function_data:
function_selector: 902
is_private: 1
is_constructor: 1

emitted_events:
[ 0xa01 0xa02 0xa03 0xa04 ]
vk_hash: 0x903
portal_contract_address: 0x404040404040404040404040404040404040404
pay_fee_from_l1: 1
pay_fee_from_public_l2: 0
called_from_l1: 1
called_from_public_l2: 0
 call_stack_item_hash: 0x902
function_data:
function_selector: 903
is_private: 1
is_constructor: 1

emitted_events:
[ 0xa02 0xa03 0xa04 0xa05 ]
vk_hash: 0x904
portal_contract_address: 0x505050505050505050505050505050505050505
pay_fee_from_l1: 1
pay_fee_from_public_l2: 0
called_from_l1: 1
called_from_public_l2: 0
 call_stack_item_hash: 0x903
function_data:
function_selector: 904
is_private: 1
is_constructor: 1

emitted_events:
[ 0xa03 0xa04 0xa05 0xa06 ]
vk_hash: 0x905
portal_contract_address: 0x606060606060606060606060606060606060606
pay_fee_from_l1: 1
pay_fee_from_public_l2: 0
called_from_l1: 1
called_from_public_l2: 0
 ]

constants:
old_tree_roots: private_data_tree_root: 0x300
nullifier_tree_root: 0x301
contract_tree_root: 0x302
private_kernel_vk_tree_root: 0x303

tx_context: is_fee_payment_tx: 0
is_rebate_payment_tx: 0
is_contract_deployment_tx: 1
contract_deployment_data: 
constructor_vk_hash: 0x304
function_tree_root: 0x305
contract_address_salt: 0x306
portal_contract_address: 0x707070707070707070707070707070707070707


is_private: 1

proof: [ 80 80 80 80 80 80 80 80 80 80 80 80 80 80 80 80 ]
vk: 0x136b80
vk_index: 42
vk_path: [ 0x1000 0x1001 0x1002 ]
 ]
start_nullifier_tree_snapshot:
root: 0x100
next_available_leaf_index: 256

low_nullifier_leaf_preimages:
[ leaf_value: 0x1000
next_value: 0x1100
next_index: 1200
 leaf_value: 0x1001
next_value: 0x1101
next_index: 1201
 leaf_value: 0x1002
next_value: 0x1102
next_index: 1202
 leaf_value: 0x1003
next_value: 0x1103
next_index: 1203
 leaf_value: 0x1004
next_value: 0x1104
next_index: 1204
 leaf_value: 0x1005
next_value: 0x1105
next_index: 1205
 leaf_value: 0x1006
next_value: 0x1106
next_index: 1206
 leaf_value: 0x1007
next_value: 0x1107
next_index: 1207
 leaf_value: 0x1008
next_value: 0x1108
next_index: 1208
 leaf_value: 0x1009
next_value: 0x1109
next_index: 1209
 leaf_value: 0x100a
next_value: 0x110a
next_index: 120a
 leaf_value: 0x100b
next_value: 0x110b
next_index: 120b
 leaf_value: 0x100c
next_value: 0x110c
next_index: 120c
 leaf_value: 0x100d
next_value: 0x110d
next_index: 120d
 leaf_value: 0x100e
next_value: 0x110e
next_index: 120e
 leaf_value: 0x100f
next_value: 0x110f
next_index: 120f
 leaf_value: 0x1010
next_value: 0x1110
next_index: 1210
 leaf_value: 0x1011
next_value: 0x1111
next_index: 1211
 leaf_value: 0x1012
next_value: 0x1112
next_index: 1212
 leaf_value: 0x1013
next_value: 0x1113
next_index: 1213
 leaf_value: 0x1014
next_value: 0x1114
next_index: 1214
 leaf_value: 0x1015
next_value: 0x1115
next_index: 1215
 leaf_value: 0x1016
next_value: 0x1116
next_index: 1216
 leaf_value: 0x1017
next_value: 0x1117
next_index: 1217
 leaf_value: 0x1018
next_value: 0x1118
next_index: 1218
 leaf_value: 0x1019
next_value: 0x1119
next_index: 1219
 leaf_value: 0x101a
next_value: 0x111a
next_index: 121a
 leaf_value: 0x101b
next_value: 0x111b
next_index: 121b
 leaf_value: 0x101c
next_value: 0x111c
next_index: 121c
 leaf_value: 0x101d
next_value: 0x111d
next_index: 121d
 leaf_value: 0x101e
next_value: 0x111e
next_index: 121e
 leaf_value: 0x101f
next_value: 0x111f
next_index: 121f
 ]
low_nullifier_membership_witness:
[ leaf_index: 2000
sibling_path: [ 0x2000 0x2001 0x2002 0x2003 0x2004 0x2005 0x2006 0x2007 ]
 leaf_index: 2001
sibling_path: [ 0x2001 0x2002 0x2003 0x2004 0x2005 0x2006 0x2007 0x2008 ]
 leaf_index: 2002
sibling_path: [ 0x2002 0x2003 0x2004 0x2005 0x2006 0x2007 0x2008 0x2009 ]
 leaf_index: 2003
sibling_path: [ 0x2003 0x2004 0x2005 0x2006 0x2007 0x2008 0x2009 0x200a ]
 leaf_index: 2004
sibling_path: [ 0x2004 0x2005 0x2006 0x2007 0x2008 0x2009 0x200a 0x200b ]
 leaf_index: 2005
sibling_path: [ 0x2005 0x2006 0x2007 0x2008 0x2009 0x200a 0x200b 0x200c ]
 leaf_index: 2006
sibling_path: [ 0x2006 0x2007 0x2008 0x2009 0x200a 0x200b 0x200c 0x200d ]
 leaf_index: 2007
sibling_path: [ 0x2007 0x2008 0x2009 0x200a 0x200b 0x200c 0x200d 0x200e ]
 leaf_index: 2008
sibling_path: [ 0x2008 0x2009 0x200a 0x200b 0x200c 0x200d 0x200e 0x200f ]
 leaf_index: 2009
sibling_path: [ 0x2009 0x200a 0x200b 0x200c 0x200d 0x200e 0x200f 0x2010 ]
 leaf_index: 200a
sibling_path: [ 0x200a 0x200b 0x200c 0x200d 0x200e 0x200f 0x2010 0x2011 ]
 leaf_index: 200b
sibling_path: [ 0x200b 0x200c 0x200d 0x200e 0x200f 0x2010 0x2011 0x2012 ]
 leaf_index: 200c
sibling_path: [ 0x200c 0x200d 0x200e 0x200f 0x2010 0x2011 0x2012 0x2013 ]
 leaf_index: 200d
sibling_path: [ 0x200d 0x200e 0x200f 0x2010 0x2011 0x2012 0x2013 0x2014 ]
 leaf_index: 200e
sibling_path: [ 0x200e 0x200f 0x2010 0x2011 0x2012 0x2013 0x2014 0x2015 ]
 leaf_index: 200f
sibling_path: [ 0x200f 0x2010 0x2011 0x2012 0x2013 0x2014 0x2015 0x2016 ]
 leaf_index: 2010
sibling_path: [ 0x2010 0x2011 0x2012 0x2013 0x2014 0x2015 0x2016 0x2017 ]
 leaf_index: 2011
sibling_path: [ 0x2011 0x2012 0x2013 0x2014 0x2015 0x2016 0x2017 0x2018 ]
 leaf_index: 2012
sibling_path: [ 0x2012 0x2013 0x2014 0x2015 0x2016 0x2017 0x2018 0x2019 ]
 leaf_index: 2013
sibling_path: [ 0x2013 0x2014 0x2015 0x2016 0x2017 0x2018 0x2019 0x201a ]
 leaf_index: 2014
sibling_path: [ 0x2014 0x2015 0x2016 0x2017 0x2018 0x2019 0x201a 0x201b ]
 leaf_index: 2015
sibling_path: [ 0x2015 0x2016 0x2017 0x2018 0x2019 0x201a 0x201b 0x201c ]
 leaf_index: 2016
sibling_path: [ 0x2016 0x2017 0x2018 0x2019 0x201a 0x201b 0x201c 0x201d ]
 leaf_index: 2017
sibling_path: [ 0x2017 0x2018 0x2019 0x201a 0x201b 0x201c 0x201d 0x201e ]
 leaf_index: 2018
sibling_path: [ 0x2018 0x2019 0x201a 0x201b 0x201c 0x201d 0x201e 0x201f ]
 leaf_index: 2019
sibling_path: [ 0x2019 0x201a 0x201b 0x201c 0x201d 0x201e 0x201f 0x2020 ]
 leaf_index: 201a
sibling_path: [ 0x201a 0x201b 0x201c 0x201d 0x201e 0x201f 0x2020 0x2021 ]
 leaf_index: 201b
sibling_path: [ 0x201b 0x201c 0x201d 0x201e 0x201f 0x2020 0x2021 0x2022 ]
 leaf_index: 201c
sibling_path: [ 0x201c 0x201d 0x201e 0x201f 0x2020 0x2021 0x2022 0x2023 ]
 leaf_index: 201d
sibling_path: [ 0x201d 0x201e 0x201f 0x2020 0x2021 0x2022 0x2023 0x2024 ]
 leaf_index: 201e
sibling_path: [ 0x201e 0x201f 0x2020 0x2021 0x2022 0x2023 0x2024 0x2025 ]
 leaf_index: 201f
sibling_path: [ 0x201f 0x2020 0x2021 0x2022 0x2023 0x2024 0x2025 0x2026 ]
 ]
historic_private_data_tree_root_membership_witnesses:
[ leaf_index: 3000
sibling_path: [ 0x3000 0x3001 0x3002 0x3003 0x3004 0x3005 0x3006 0x3007 ]
 leaf_index: 4000
sibling_path: [ 0x4000 0x4001 0x4002 0x4003 0x4004 0x4005 0x4006 0x4007 ]
 ]
historic_contract_tree_root_membership_witnesses:
[ leaf_index: 5000
sibling_path: [ 0x5000 0x5001 0x5002 0x5003 0x5004 0x5005 0x5006 0x5007 ]
 leaf_index: 6000
sibling_path: [ 0x6000 0x6001 0x6002 0x6003 0x6004 0x6005 0x6006 0x6007 ]
 ]
constants:
start_tree_of_historic_private_data_tree_roots_snapshot:
 root: 0x100
next_available_leaf_index: 256

start_tree_of_historic_contract_tree_roots_snapshot:
root: 0x200
next_available_leaf_index: 512

tree_of_historic_l1_to_l2_msg_tree_roots_snapshot:
root: 0x300
next_available_leaf_index: 768

private_kernel_vk_tree_root: 0x401
public_kernel_vk_tree_root: 0x402
base_rollup_vk_hash: 0x403
merge_rollup_vk_hash: 0x404

prover_id: 0x42
"
`;

exports[`structs/base_rollup serializes and prints BaseRollupPublicInputs 1`] = `
"rollup_type: 0
end_aggregation_object:
P0: { 0x100, 0x101 }
P1: { 0x200, 0x201 }
public_inputs: [
  0: 0x102
  1: 0x103
  2: 0x104
  3: 0x105
]
proof_witness_indices: [
  0: 262
  1: 263
  2: 264
  3: 265
  4: 266
  5: 267
has_data: 0

constants:
start_tree_of_historic_private_data_tree_roots_snapshot:
 root: 0x200
next_available_leaf_index: 512

start_tree_of_historic_contract_tree_roots_snapshot:
root: 0x300
next_available_leaf_index: 768

tree_of_historic_l1_to_l2_msg_tree_roots_snapshot:
root: 0x400
next_available_leaf_index: 1024

private_kernel_vk_tree_root: 0x501
public_kernel_vk_tree_root: 0x502
base_rollup_vk_hash: 0x503
merge_rollup_vk_hash: 0x504

start_nullifier_tree_snapshot:
root: 0x300
next_available_leaf_index: 768

end_nullifier_tree_snapshots:
root: 0x400
next_available_leaf_index: 1024

new_commitments_subtree_root: 0x501
new_nullifiers_subtree_root: 0x502
new_contract_leaves_subtree_root: 0x503
new_commitments_hash: 0x601
new_nullifiers_hash: 0x602
new_l1_msgs_hash: 0x603
new_contract_data_hash: 0x604
prover_contributions_hash: 0x605
"
`;
